[WIP] cmd: experiment (#1926)
iterative/dvc.org: content/docs/command-reference/exp/apply.md

> # exp apply

Reviewer: All refs should use the full name probably, e.g.:
```suggestion
# expperiment apply
```

Reviewer: Oops, typo (fixed).

> Apply the changes from a previously run experiment into the workspace.

Reviewer: ```suggestion
Apply the changes from a previously run experiment into the
<abbr>workspace</abbr>.
```

> Apply the changes from a previously run experiment into the workspace.

Reviewer: ```suggestion
Apply the changes from a previous experiment to the
<abbr>workspace</abbr>.
```
> "into" does sound better for some reason... but I think the right phrase is "apply to" per https://hinative.com/en-US/questions/1240152

Reviewer: Maybe link "experiment" to the index for now (we can link to some concept page or guide later).

## Synopsis

```usage
usage: dvc exp apply [-h] [-q | -v] experiment

positional arguments:

>   experiment     Experiment to be applied.

Reviewer: Core question: Should `experiment` argument be `target` instead for consistency with existing commands?

Author: `experiment` here is an experiment name or git SHA, equivalent to a git branch/tag name or git SHA. Not sure what the ideal usage would be here, but `target` seems like it might be confusing with the `dvc repro`/`exp run` `target`?

Reviewer: Currently each command gives a different meaning to `target(s)` anyway. It varies a lot, it can be files, stages, .dvc files, dvc.yaml files, directories... (e.g. add, check-ignore, commit, freeze, repro)...

Reviewer: But yeah I guess that's not ideal to begin with. And they do tend to be DVC-tracked stuff so I take this back. Resolving

Reviewer: If anything I'd consider shortening the arg name to `exp`.

```

## Description

> This command will apply the changes from an experiment to the workspace. It is

Reviewer: "from an existing experiment"?
"previous experiment"?
"experiment in {some term for the log or cache of exps}"?

Reviewer: Also you can just start descriptions with the verb in 2nd person e.g. "Applies the changes..." — we tend to do that in cmd ref descs.

> usually run when you are ready to Git commit the results of an experiment into
> the project.

Reviewer: ```suggestion
usually run when you are ready to commit the results of an experiment with
Git.
```

Reviewer: And so do all experiment commands require the project to use Git? May be something to note prominently somewhere.

> This command will fail if the specified experiment was not derived from the Git
> commit which is currently (Git) checkout-ed in the workspace.

Reviewer: ```suggestion
This command will fail if the target `experiment` was not derived from the
current Git commit.
```

## Options

> 
> - `experiment` - name or revision of the experiment to be checked out.

Reviewer: No need to include here for now unless there's very complex logic about this argument. We will revisit this question later for all refs in #2024
```suggestion
```

Reviewer: We should explain the ways to refer to experiments (name or revision) somewhere. If that's in the Description then this isn't needed (I incline for that), otherwise it can be here.

```

We can inspect these changes with Git:

```dvc
$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
        modified:   dvc.lock
        modified:   params.yaml
        modified:   prc.json
        modified:   scores.json

$ git diff params.yaml scores.json
diff --git a/params.yaml b/params.yaml
index 4c4d898..faf781a 100644
--- a/params.yaml
+++ b/params.yaml
@@ -3,7 +3,7 @@ prepare:
   seed: 20170428

 featurize:
-  max_features: 1500
+  max_features: 2000
   ngrams: 2

 train:
diff --git a/scores.json b/scores.json
index c995f24..c640c4e 100644
--- a/scores.json
+++ b/scores.json
@@ -1 +1 @@
-{"auc": 0.6131382960762474}
\ No newline at end of file
+{"auc": 0.5775633054725381}
\ No newline at end of file
```

and with DVC:

```
$ dvc status
Data and pipelines are up to date.

$ dvc diff
Modified:
    data/features/
    data/features/test.pkl
    data/features/train.pkl
    model.pkl
    prc.json
    scores.json

files summary: 0 added, 0 deleted, 5 modified, 0 not in cache
```

To promote this experiment we simply `git add` and `git commit` the changes:

```dvc
$ git add .

> $ git commit -m "promote experiment exp-1dad0"

Reviewer: Wow this example is great, thanks @pmrowla ! I'd try to remove some console output if it's not directly related to `apply` (to make the doc easier to maintain in the future) and the same for this last part about `exp branch`: not sure we want it in this ref.

But if you have some complex examples that combine several exp subcommands I'd put those in the index for now.

Reviewer: p.s. it's OK if experiments or parts are repeated across refs.
